### ISSUE TITLE
DNS plugins don't work if _acme-challenge is a separate zone

### ISSUE THREAD

**Issue by patrakov** [certbot/certbot#7701]
## My operating system is (include version):
Debian buster

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):
pip

## What's unusual in my setup:

`smb.mydomain.com` is the domain for which I want to get a wildcard. However, my publicly available authoritative PowerDNS server is configured as a slave of a different server that I would rather not touch with certbot. Instead, a separate master zone, `_acme-challenge.smb.mydomain.com`, is configured on PowerDNS.

## I ran this command and it produced this output:
```
certbot ... --non-interactive --authenticator certbot-dns-powerdns:dns-powerdns --certbot-dns-powerdns:dns-powerdns-credentials /etc/letsencrypt/pdns-credentials.ini --expand --domains smb.mydomain.com,*.smb.mydomain.com
```

```
...
2020-01-21 16:45:29,232:DEBUG:urllib3.connectionpool:http://172.31.100.21:8081 "GET /api/v1/servers/localhost/zones/smb.mydomain.com. HTTP/1.1" 200 4094
...
2020-01-21 16:45:29,255:DEBUG:urllib3.connectionpool:http://172.31.100.21:8081 "PATCH /api/v1/servers/localhost/zones/smb.mydomain.com. HTTP/1.1" 204 0
...
Domain: smb.mydomain.com
Type:   unauthorized
Detail: No TXT record found at _acme-challenge.smb.mydomain.com
```

## Certbot's behavior differed from what I expected because:

It made the edit to the wrong zone. It created an `_acme-challenge` TXT record in the `smb.mydomain.com` slave zone, but should have created an `@` TXT record in the `_acme-challenge.smb.mydomain.com` zone.

## Workaround

I have edited the dns_common_lexicon.py file. Replaced this line:

        domain_name_guesses = dns_common.base_domain_name_guesses(domain)

with this one:

        domain_name_guesses = dns_common.base_domain_name_guesses('_acme-challenge.' + domain)

After that, it works as expected.

**Comment by bmw** [comment]
@adferrand or @joohoi, can you comment on this? I'm personally not immediately sure what is the best path forward here.

**Comment by joohoi** [comment]
I think the workaround is good, and handles the use case of pointing the validation domain apex to a server that the user would be comfortable storing the credentials to pretty well. 

Just to note that in order to implement this, we should make sure that all of our DNS plugins behave consistently here - the underlying issue this would fix is not provider specific.

**Comment by adferrand** [comment]
For an immediate mitigation, I think also it is the best approach. The idea here is to try the Lexicon provider `authenticate()` method from the most specific domain to the less specific one (eg. `['www.example.com', 'example.com', 'com']` for `www.example.com`), and stop the loop once the `authenticate()` method does not a raise an exception anymore: at this point we can state that the current subdomain is the right one, since credentials and provider given by the user are correct for this subdomain.

This behavior relies on `authenticate()` on Lexicon side to consistently raises an exception when the authentication is invalid.

Adding `_acme.www.example.com` at the top of the `['www.example.com', 'example.com', 'com']` list does not change any logic on the loop, so relies on the same behavior and so should work out of the box on all existing providers that are already implemented on Certbot. I can not make an promise for the other providers, but since they are not available on Certbot, it is not a problem.

**Comment by bmw** [comment]
That makes a lot of sense. Thanks for the responses.

Unless someone can think of a case where this causes existing setups to break, I think we'd accept a PR with this change.

**Comment by patrakov** [comment]
Well,  I agree that we should make sure that all of our DNS plugins behave consistently here. The same `dns_common.base_domain_name_guesses` function is also called in `dns_cloudflare`, `dns_digitalocean`, `dns_google` and `dns_rfc2136` plugins. At least in the first three cases, it looks like the same problem exists, but I cannot confirm it definitely as of now. I don't have a Cloudflare or DigitalOcean account, and Google does not let me use their cloud DNS because I do not represent an organization.

The `dns_rfc2136` plugin does the guess from within its `add_txt_record` function, based on the record name only (i.e. does not use the bare domain) and therefore is not affected.

Do I understand correctly that the pull request should change not only dns_lexicon_base.py, but also `dns_digitalocean`, `dns_google` and `dns_rfc2136` plugins? If so, I would have to rely on someone else to validate it.

**Comment by adferrand** [comment]
Yes I guess that this issue can happen also on non Lexicon-based plugins. If the PR fixes the approach also for these plugins, I can personally validate for cloudflare and digitalocean. I think @joohoi has already working setups of bind9 to test rfc2136. For google, I found a way to have a temporary free account to test their DNS when I implemented the Google plugin on Lexicon, so I think there is a way to test it also.
